lib/srv/reexec.go

```diff
@@ -764,39 +607,208 @@ func runForward(handler forwardHandler) (errw io.Writer, code int, err error) {
 	for {
 		buf := make([]byte, 1024)
 		fbuf := make([]*os.File, 1)
-		n, fn, err := uds.ReadWithFDs(conn, buf, fbuf)
+		n, fn, err := uds.ReadWithFDs(parentConn, buf, fbuf)
 		if err != nil {
 			if utils.IsOKNetworkError(err) {
 				return errorWriter, teleport.RemoteCommandSuccess, nil
 			}
 			return errorWriter, teleport.RemoteCommandFailure, trace.Wrap(err)
 		}
-		addr := string(buf[:n])
-		if fn == 0 {
-			log.Errorf("Parent did not send a file descriptor for address %q.", addr)
-			continue
-		}
 
 		go func() {
-			if err := handler(ctx, addr, fbuf[0]); err != nil {
-				log.WithError(err).Errorf("Error handling forwarding request for address %q.", addr)
+			var req networking.Request
+			if err := json.Unmarshal(buf[:n], &req); err != nil {
+				slog.With("error", err).ErrorContext(ctx, "Error parsing networking request.")
+				return
+			}
+			log := slog.With("request", req)
+
+			if fn == 0 {
+				log.ErrorContext(ctx, "Networking request requires a control file.")
+				return
+			}
+
+			controlConn, err := uds.FromFile(fbuf[0])
+			_ = fbuf[0].Close()
+			if err != nil {
+				log.ErrorContext(ctx, "Failed to get a connection from control file.")
+				return
+			}
+			defer controlConn.Close()
+
+			log.Debug("Handling networking request")
+			netFile, err := handleNetworkingRequest(ctx, req)
+			if err != nil {
+				log.With("error", err).ErrorContext(ctx, "Error handling networking request.")
+				if _, err2 := controlConn.Write([]byte(err.Error())); err2 != nil {
+					log.With("error", err2.Error()).ErrorContext(ctx, "Failed to write error to control conn.")
+				}
+			}
+			defer netFile.Close()
+
+			if _, _, err := uds.WriteWithFDs(controlConn, nil, []*os.File{netFile}); err != nil {
+				log.With("error", err.Error()).ErrorContext(ctx, "Failed to write networking file to control conn.")
 			}
 		}()
 	}
 }
 
-// RunLocalForward reads in the command to run from the parent process (over a
-// pipe) then port forwards.
-func RunLocalForward() (errw io.Writer, code int, err error) {
-	errw, code, err = runForward(handleLocalPortForward)
-	return errw, code, trace.Wrap(err)
+func handleNetworkingRequest(ctx context.Context, req networking.Request) (*os.File, error) {
+	ctx, cancel := context.WithTimeout(ctx, 30*time.Second)
+	defer cancel()
+
+	switch req.Operation {
+	case networking.NetworkingOperationDial:
+		var d net.Dialer
+		conn, err := d.DialContext(ctx, req.Network, req.Address)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		defer conn.Close()
+
+		connFD, err := getConnFile(conn)
+		return connFD, trace.Wrap(err)
+
+	case networking.NetworkingOperationListen:
```

Comment:
Should we add an operation for listenpacket too?

Reply:
We don't have a use for it yet, do we?